src/sentry/digests/utilities.py

```diff
@@ -23,3 +24,134 @@ def get_digest_metadata(digest):
                     end = record.datetime
 
     return start, end, counts
+
+
+def get_personalized_digests(project_id, digest, user_ids):
+    """
+    get_personalized_digests(project: Project, digest: Digest, users: Set[User]) -> Iterator[User, Digest]
+    """
+    # TODO(LB): I Know this is inefficent.
+    # In the case that ProjectOwnership does exist, I do the same query twice.
+    # Once with this statement and again with the call to ProjectOwnership.get_actors()
+    # Will follow up with another PR to reduce the number of queries.
+    if ProjectOwnership.objects.filter(project_id=project_id).exists():
+        events = get_events_from_digest(digest)
+        events_by_actor = build_events_by_actor(project_id, events)
+        events_by_user = convert_actors_to_user_set(events_by_actor, user_ids)
+        for user_id in user_ids:
+            yield user_id, build_custom_digest(digest, events_by_user[user_id])
+    else:
+        for user_id in user_ids:
+            yield user_id, digest
+
+
+def sort_records(records):
+    """
+    Sorts records for fetch_state method
+    fetch_state is expecting these records to be ordered from newest to oldest
+    """
+    return sorted(records, key=lambda r: r.value.event.datetime, reverse=True)
+
+
+def get_events_from_digest(digest):
+    events = []
+    for rule_groups in six.itervalues(digest):
+        for group_records in six.itervalues(rule_groups):
+            events.append(group_records[0].value.event)
+    return set(events)
+
+
+def build_custom_digest(original_digest, events):
+    """
+    build_custom_digest(original_digest: Digest, user_id: Int, events_by_users: Map[User_Id:Set(Events)]) -> Digest
+    """
+    user_digest = OrderedDict()
+    for rule, rule_groups in six.iteritems(original_digest):
+        user_rule_groups = OrderedDict()
+        for group, group_records in six.iteritems(rule_groups):
+            user_group_records = [
+                record for record in group_records
+                if record.value.event in events
+            ]
+            if user_group_records:
+                user_rule_groups[group] = sort_records(user_group_records)
+        if user_rule_groups:
+            user_digest[rule] = user_rule_groups
+    return user_digest
+
+
+def build_events_by_actor(project_id, events):
+    """
+    build_events_by_actor(project_id: Int, events: Set(Events)) -> Map[Actor:Set(Events)]
+    """
+    events_by_actor = {}
+    for event in events:
+        # TODO(LB): I Know this is inefficent.
+        # ProjectOwnership.get_owners is O(n) queries and I'm doing that O(len(events)) times
+        # I will create a follow-up PR to address this method's efficency problem
+        # Just wanted to make as few changes as possible for now.
+        actors, __ = ProjectOwnership.get_owners(project_id, event.data)
+        if actors == ProjectOwnership.Everyone:
+            actors = [actors]
```

Comment:
Eh... I liked this way to isolate all the converting to user_ids in one method. As opposed to doing it twice in two different methods. I view Everyone as an Actor. Just like a Team is an actor. A team is a collection of user_ids and so is Everyone. It's just a team that spans all sendable users. But if you understand that differently I understand. I just disagree here

Reply:
I disagree pretty strongly with this (it doesn't fulfill the type signature implied by the function docstring since `Everyone` is not generically substitutable for an `Actor`) but I'm not going to block the review on it.

Reply:
@tkaemming ok then. I can restructure it.